vm-virtio/src/console.rs

```diff
@@ -0,0 +1,353 @@
+            epoll::Event::new(epoll::Events::EPOLLIN, u64::from(INPUT_QUEUE_EVENT)),
+        )
+        .map_err(DeviceError::EpollCtl)?;
+        epoll::ctl(
+            epoll_fd,
+            epoll::ControlOptions::EPOLL_CTL_ADD,
+            self.output_queue_evt.as_raw_fd(),
+            epoll::Event::new(epoll::Events::EPOLLIN, u64::from(OUTPUT_QUEUE_EVENT)),
+        )
+        .map_err(DeviceError::EpollCtl)?;
+       epoll::ctl(
+            epoll_fd,
+            epoll::ControlOptions::EPOLL_CTL_ADD,
+            self.kill_evt.as_raw_fd(),
+            epoll::Event::new(epoll::Events::EPOLLIN, u64::from(KILL_EVENT)),
+        )
+        .map_err(DeviceError::EpollCtl)?;
+
+        const EPOLL_EVENTS_LEN: usize = 100;
+        let mut events = vec![epoll::Event::new(epoll::Events::empty(), 0); EPOLL_EVENTS_LEN];
+
+        'epoll: loop {
+            let num_events =
+                epoll::wait(epoll_fd, -1, &mut events[..]).map_err(DeviceError::EpollWait)?;
+
+            for event in events.iter().take(num_events) {
+                let ev_type = event.data as u16;
+
+                match ev_type {
+                    INPUT_QUEUE_EVENT => {
+                        if let Err(e) = self.input_queue_evt.read() {
+                            error!("Failed to get queue event: {:?}", e);
+                            break 'epoll;
+                        } else if self.process_input_queue() {
+                            if let Err(e) = self.signal_used_queue() {
+                                error!("Failed to signal used queue: {:?}", e);
+                                break 'epoll;
+                            }
+                        }
+                    }
+                    OUTPUT_QUEUE_EVENT => {
+                        if let Err(e) = self.output_queue_evt.read() {
+                            error!("Failed to get queue event: {:?}", e);
+                            break 'epoll;
+                        } else {
+				self.process_output_queue();
+			}
+                    }
+                    KILL_EVENT => {
+                        debug!("KILL_EVENT received, stopping epoll loop");
+                        break 'epoll;
+                    }
+                    _ => {
+                        error!("Unknown event for virtio-block");
+                    }
+                }
+            }
+        }
+
+        Ok(())
+    }
+}
+
+/// Virtio device for exposing console to the guest OS through virtio.
+pub struct Console{
+    kill_evt: Option<EventFd>,
+    input_queue_evt: Option<EventFd>,
+    avail_features: u64,
+    acked_features: u64,
+    in_buffer: Arc<Mutex<VecDeque<u8>>>,
+    out: Option<Box<io::Write + Send>>,
+}
+
+impl Console{
+    /// Create a new virtio console device that gets random data from /dev/urandom.
+    pub fn new(out: Option<Box<io::Write + Send>>) -> io::Result<Console> {
+        let avail_features = 1u64 << VIRTIO_F_VERSION_1;// | (1u64 << VIRTIO_CONSOLE_F_SIZE);
```

Comment:
Commented out feature?

Reply:
Yes. Since it is the simplest version, I decided to add this feature in the next iteration but still kept it here as a reminder. I can remove it if doesn't look good.